### ISSUE TITLE
nifi-config locates controller service dependencies by GUID rather than by name

### ISSUE THREAD

**Issue by ennui93** [hermannpencole/nifi-config#35]
At present, as nifi-config captures inter-dependencies between controller services as GUIDs rather than capturing them by name, it is not possible to update the configuration of controller services where one controller service depends on another.

# Steps
1. Import the [controller_services.xml](https://gist.github.com/ennui93/9bac197de1e43b3adbb451b93e255f7e#file-controller_services-xml) template into a given NiFi instance, which creates three controller services: `CSVReader`, `AvroSchemaRegistry`, and `JsonRecordSetWriter`. Note that `JsonRecordSetWriter` has a dependency on  `AvroSchemaRegistry`.
This can be done either manually through the NiFi web canvas UI, or with nifi-config, à la:
`java -jar nifi-config/target/nifi-deploy-config-*.jar -m deployTemplate -b "root" -c controller_services.xml -n http://localhost:8079/nifi-api`
2. Update the Schema Registry property in the `JsonRecordSetWriter` controller service to point to the `AvroSchemaRegistry` controller service.
(At present, nifi-config stores controller service inter-dependencies in its configuration by GUID rather than by name (see [the value of the schema-registry property](https://gist.github.com/ennui93/9bac197de1e43b3adbb451b93e255f7e#file-controller_services-json-L8)) so this is only possible through the NiFi web canvas UI. This is itself a manifestation of the issue described herein.)
3. Enable all three controller services. At this point, we have simulated existing controller services in an arbitrary environment.
4. Import the [controller_service_dependencies.xml](https://gist.github.com/ennui93/9bac197de1e43b3adbb451b93e255f7e#file-controller_service_dependencies-xml) template into a process group under the root node on the given NiFi instance:
`java -jar nifi-config/target/nifi-deploy-config-*.jar -m deployTemplate -b "root > controller_service_dependencies" -c controller_service_dependencies.xml -n http://localhost:8079/nifi-api`
5. Attempt to update the configuration on the newly instantiated template in "root > controller_service_dependencies" using [controller_service_dependencies.json](https://gist.github.com/ennui93/9bac197de1e43b3adbb451b93e255f7e#file-controller_service_dependencies-json)

# Expected result
nifi-config locates the `AvroSchemaRegistry` controller service by name and updates the `JsonRecordSetWriter`'s `Schema Registry` property appropriately.

# Actual result
```
2017-12-21 16:11:14 | INFO  | [main] c.g.h.n.c.Main:109 - Starting config_nifi 1.1.27-SNAPSHOT on mode updateConfig
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.Main:131 - Communicate with nifi 1.3.0
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:65 - Processing : controller_service_dependencies.json
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.ProcessorService:54 -  GetFile (33b7ba02-6f96-3915-fa0e-56d32b97df9a) is already 
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.ProcessorService:54 -  ConvertRecord (55f2cf94-a952-343a-705d-579c12bdedae) is already 
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:75 - [root, controller_service_dependencies] is stopped
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:239 - Update processor : ConvertRecord
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:271 - Updated : ConvertRecord
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:239 - Update processor : GetFile
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:271 - Updated : GetFile
2017-12-21 16:11:15 | INFO  | [pool-1-thread-1] c.g.h.n.c.s.ControllerServicesService:100 - 3efc8893-ab75-35d9-57d4-513160854046 is DISABLED
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:239 - Update processor : ConvertRecord
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:271 - Updated : ConvertRecord
2017-12-21 16:11:15 | INFO  | [main] c.g.h.n.c.s.ProcessorService:54 -  ConvertRecord (55f2cf94-a952-343a-705d-579c12bdedae) is already 
2017-12-21 16:11:15 | INFO  | [pool-4-thread-1] c.g.h.n.c.s.ControllerServicesService:100 - 3efc8893-ab75-35d9-6108-ef59be5033bf is DISABLED
2017-12-21 16:11:15 | ERROR | [main] c.g.h.n.c.Main:154 - Not Found
com.github.hermannpencole.nifi.swagger.ApiException: Not Found
	at com.github.hermannpencole.nifi.swagger.ApiClient.handleResponse(ApiClient.java:1104)
	at com.github.hermannpencole.nifi.swagger.ApiClient.execute(ApiClient.java:1027)
	at com.github.hermannpencole.nifi.swagger.client.ControllerServicesApi.updateControllerServiceWithHttpInfo(ControllerServicesApi.java:884)
	at com.github.hermannpencole.nifi.swagger.client.ControllerServicesApi.updateControllerService(ControllerServicesApi.java:869)
	at com.github.hermannpencole.nifi.config.service.ControllerServicesService.updateControllerService(ControllerServicesService.java:66)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateControllers(UpdateProcessorService.java:156)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateByBranch(UpdateProcessorService.java:87)
	at com.github.hermannpencole.nifi.config.Main.main(Main.java:136)
Exception in thread "main" com.github.hermannpencole.nifi.config.model.ConfigException: Not Found: Unable to locate controller service with id '8d17b40b-5cd3-3134-6dbb-7f27b164c785'.
	at com.github.hermannpencole.nifi.config.Main.main(Main.java:155)
Caused by: com.github.hermannpencole.nifi.swagger.ApiException: Not Found
	at com.github.hermannpencole.nifi.swagger.ApiClient.handleResponse(ApiClient.java:1104)
	at com.github.hermannpencole.nifi.swagger.ApiClient.execute(ApiClient.java:1027)
	at com.github.hermannpencole.nifi.swagger.client.ControllerServicesApi.updateControllerServiceWithHttpInfo(ControllerServicesApi.java:884)
	at com.github.hermannpencole.nifi.swagger.client.ControllerServicesApi.updateControllerService(ControllerServicesApi.java:869)
	at com.github.hermannpencole.nifi.config.service.ControllerServicesService.updateControllerService(ControllerServicesService.java:66)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateControllers(UpdateProcessorService.java:156)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateByBranch(UpdateProcessorService.java:87)
	at com.github.hermannpencole.nifi.config.Main.main(Main.java:136)
```

# Workaround
Replace the GUID `8d17b40b-5cd3-3134-6dbb-7f27b164c785` appearing in the `JsonRecordSetWriter` controller service's `schema-registry` property value with the newly generated GUID for the `AvroSchemaRegistry` before running the `updateConfig` step.

**Comment by hermannpencole** [comment]
I have pushed a correction, maybe it will work
Merry christmas
I will make a release next week

**Comment by ennui93** [comment]
As of 1d78c61658a59072729063aabd5ffdc0f626e8b3, attempting to follow the above steps results in the failure below, where `3efc8893-ab75-35d9-e2a9-24d8e6bc27d2` is the GUID of the deleted controller service declaration in the `controller_service_dependencies` child group:
```
$ java -jar nifi-config/target/nifi-deploy-config-*.jar -m updateConfig -b "root > controller_service_dependencies" -c controller_service_dependencies.json -n http://localhost:8079/nifi-api
2017-12-27 11:29:14 | INFO  | [main] c.g.h.n.c.Main:109 - Starting config_nifi 1.1.29-SNAPSHOT on mode updateConfig

[…]

2017-12-27 11:29:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:261 - Update config processor 55f2cf94-a952-343a-e194-5c37c54d18b7 ({}) ConvertRecord
2017-12-27 11:29:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:293 - Processor 55f2cf94-a952-343a-e194-5c37c54d18b7 ({}) have config updated ConvertRecord
2017-12-27 11:29:15 | INFO  | [main] c.g.h.n.c.s.ControllerServicesService:106 -  JsonRecordSetWriter (3efc8893-ab75-35d9-e2a9-24d8e6bc27d2) update for DISABLED
2017-12-27 11:29:15 | INFO  | [pool-1-thread-1] c.g.h.n.c.s.ControllerServicesService:110 -  JsonRecordSetWriter (3efc8893-ab75-35d9-e2a9-24d8e6bc27d2) waiting for DISABLED
2017-12-27 11:29:15 | INFO  | [pool-1-thread-1] c.g.h.n.c.s.ControllerServicesService:112 -  JsonRecordSetWriter (3efc8893-ab75-35d9-e2a9-24d8e6bc27d2) is DISABLED
2017-12-27 11:29:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:261 - Update config processor 55f2cf94-a952-343a-e194-5c37c54d18b7 ({}) ConvertRecord
2017-12-27 11:29:15 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:293 - Processor 55f2cf94-a952-343a-e194-5c37c54d18b7 ({}) have config updated ConvertRecord
2017-12-27 11:29:15 | INFO  | [main] c.g.h.n.c.s.ControllerServicesService:106 -  JsonRecordSetWriter (3efc8893-ab75-35d9-6108-ef59be5033bf) update for DISABLED
2017-12-27 11:29:15 | INFO  | [pool-3-thread-1] c.g.h.n.c.s.ControllerServicesService:110 -  JsonRecordSetWriter (3efc8893-ab75-35d9-6108-ef59be5033bf) waiting for DISABLED
2017-12-27 11:29:15 | INFO  | [pool-3-thread-1] c.g.h.n.c.s.ControllerServicesService:112 -  JsonRecordSetWriter (3efc8893-ab75-35d9-6108-ef59be5033bf) is DISABLED
2017-12-27 11:29:15 | INFO  | [main] c.g.h.n.c.s.ControllerServicesService:76 -  JsonRecordSetWriter (3efc8893-ab75-35d9-6108-ef59be5033bf) is UPDATED
2017-12-27 11:29:15 | INFO  | [main] c.g.h.n.c.s.ControllerServicesService:106 -  CSVReader (7db4fe70-133f-3576-364a-a1a1aaf1f960) update for DISABLED
2017-12-27 11:29:15 | INFO  | [pool-4-thread-1] c.g.h.n.c.s.ControllerServicesService:110 -  CSVReader (7db4fe70-133f-3576-364a-a1a1aaf1f960) waiting for DISABLED
2017-12-27 11:29:15 | INFO  | [pool-4-thread-1] c.g.h.n.c.s.ControllerServicesService:112 -  CSVReader (7db4fe70-133f-3576-364a-a1a1aaf1f960) is DISABLED
2017-12-27 11:29:16 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:261 - Update config processor 55f2cf94-a952-343a-e194-5c37c54d18b7 ({}) ConvertRecord
2017-12-27 11:29:16 | INFO  | [main] c.g.h.n.c.s.UpdateProcessorService:293 - Processor 55f2cf94-a952-343a-e194-5c37c54d18b7 ({}) have config updated ConvertRecord
2017-12-27 11:29:16 | INFO  | [main] c.g.h.n.c.s.ControllerServicesService:106 -  CSVReader (7db4fe70-133f-3576-e00d-f31908df581b) update for DISABLED
2017-12-27 11:29:16 | INFO  | [pool-6-thread-1] c.g.h.n.c.s.ControllerServicesService:110 -  CSVReader (7db4fe70-133f-3576-e00d-f31908df581b) waiting for DISABLED
2017-12-27 11:29:16 | INFO  | [pool-6-thread-1] c.g.h.n.c.s.ControllerServicesService:112 -  CSVReader (7db4fe70-133f-3576-e00d-f31908df581b) is DISABLED
2017-12-27 11:29:16 | INFO  | [main] c.g.h.n.c.s.ControllerServicesService:76 -  CSVReader (7db4fe70-133f-3576-e00d-f31908df581b) is UPDATED
2017-12-27 11:29:16 | INFO  | [main] c.g.h.n.c.s.ControllerServicesService:106 -  AvroSchemaRegistry (8d17b40b-5cd3-3134-9b2a-7a54ad500b2c) update for DISABLED
2017-12-27 11:29:16 | INFO  | [pool-7-thread-1] c.g.h.n.c.s.ControllerServicesService:110 -  AvroSchemaRegistry (8d17b40b-5cd3-3134-9b2a-7a54ad500b2c) waiting for DISABLED
2017-12-27 11:29:16 | INFO  | [pool-7-thread-1] c.g.h.n.c.s.ControllerServicesService:112 -  AvroSchemaRegistry (8d17b40b-5cd3-3134-9b2a-7a54ad500b2c) is DISABLED
2017-12-27 11:29:16 | ERROR | [main] c.g.h.n.c.Main:154 - Not Found
com.github.hermannpencole.nifi.swagger.ApiException: Not Found
	at com.github.hermannpencole.nifi.swagger.ApiClient.handleResponse(ApiClient.java:1104)
	at com.github.hermannpencole.nifi.swagger.ApiClient.execute(ApiClient.java:1027)
	at com.github.hermannpencole.nifi.swagger.client.ControllerServicesApi.getControllerServiceWithHttpInfo(ControllerServicesApi.java:262)
	at com.github.hermannpencole.nifi.swagger.client.ControllerServicesApi.getControllerService(ControllerServicesApi.java:248)
	at com.github.hermannpencole.nifi.config.service.ControllerServicesService.getControllerServices(ControllerServicesService.java:119)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateOldReference(UpdateProcessorService.java:196)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateControllers(UpdateProcessorService.java:148)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateByBranch(UpdateProcessorService.java:88)
	at com.github.hermannpencole.nifi.config.Main.main(Main.java:136)
Exception in thread "main" com.github.hermannpencole.nifi.config.model.ConfigException: Not Found: Unable to locate controller service with id '3efc8893-ab75-35d9-e2a9-24d8e6bc27d2'.
	at com.github.hermannpencole.nifi.config.Main.main(Main.java:155)
Caused by: com.github.hermannpencole.nifi.swagger.ApiException: Not Found
	at com.github.hermannpencole.nifi.swagger.ApiClient.handleResponse(ApiClient.java:1104)
	at com.github.hermannpencole.nifi.swagger.ApiClient.execute(ApiClient.java:1027)
	at com.github.hermannpencole.nifi.swagger.client.ControllerServicesApi.getControllerServiceWithHttpInfo(ControllerServicesApi.java:262)
	at com.github.hermannpencole.nifi.swagger.client.ControllerServicesApi.getControllerService(ControllerServicesApi.java:248)
	at com.github.hermannpencole.nifi.config.service.ControllerServicesService.getControllerServices(ControllerServicesService.java:119)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateOldReference(UpdateProcessorService.java:196)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateControllers(UpdateProcessorService.java:148)
	at com.github.hermannpencole.nifi.config.service.UpdateProcessorService.updateByBranch(UpdateProcessorService.java:88)
	at com.github.hermannpencole.nifi.config.Main.main(Main.java:136)
```

**Comment by hermannpencole** [comment]
okay,
corrected and tested on the new version 1.1.30

**Comment by ennui93** [comment]
Fantastic! Thanks!
